### PR TITLE
Add primitive dist_diag. 

### PR DESCRIPTION
**Pull request by NanmiaoWu**
It is the distributed version of Numpy diag for 1-D array as input. Remove diag0d to comply with Numpy.

### CODE REVIEW

File: src/plugins/dist_matrixops/dist_diag.cpp [hunk 1]
```diff
@@ -0,0 +1,576 @@
...
+#include <phylanx/util/distributed_matrix.hpp>
+#include <phylanx/util/distributed_vector.hpp>
+
+#include <hpx/assertion.hpp>
+#include <hpx/errors/throw_exception.hpp>
+#include <hpx/include/lcos.hpp>
+#include <hpx/include/naming.hpp>
+#include <hpx/include/util.hpp>
+
+#include <atomic>
```
Comment (taless474) [inline src/plugins/dist_matrixops/dist_diag.cpp:29]: I think atomic is not needed

File: src/plugins/dist_matrixops/dist_diag.cpp [hunk 2]
```diff
@@ -0,0 +1,576 @@
...
+#include <phylanx/execution_tree/localities_annotation.hpp>
+#include <phylanx/execution_tree/locality_annotation.hpp>
+#include <phylanx/execution_tree/meta_annotation.hpp>
+#include <phylanx/execution_tree/primitives/annotate_primitive.hpp>
+#include <phylanx/execution_tree/primitives/node_data_helpers.hpp>
+#include <phylanx/execution_tree/tiling_annotations.hpp>
+#include <phylanx/ir/node_data.hpp>
+#include <phylanx/plugins/dist_matrixops/dist_diag.hpp>
+#include <phylanx/plugins/dist_matrixops/tile_calculation_helper.hpp>
+#include <phylanx/util/detail/range_dimension.hpp>
```
Comment (taless474) [inline src/plugins/dist_matrixops/dist_diag.cpp:19]: I believe range_dimension is not used

File: src/plugins/dist_matrixops/dist_diag.cpp [hunk 3]
```diff
@@ -0,0 +1,576 @@
...
+        execution_tree::primitive_arguments_type&& operands,
+        std::string const& name, std::string const& codename)
+      : primitive_component_base(std::move(operands), name, codename)
+    {}
+
+    ///////////////////////////////////////////////////////////////////////////
+
+    namespace detail
+    {
+        //static std::atomic<std::size_t> diag_count(0);
```
Comment (taless474) [inline src/plugins/dist_matrixops/dist_diag.cpp:88]: I think it would be best if you put these functions in an hpp file, so both retile and diag access it.

File: src/plugins/dist_matrixops/dist_diag.cpp [hunk 4]
```diff
@@ -0,0 +1,576 @@
...
+            tiling_span(column_start, column_start + column_size));
+
+        auto locality_ann = arr_localities.locality_.as_annotation();
+        auto attached_annotation =
+            std::make_shared<annotation>(localities_annotation(locality_ann,
+                tile_info.as_annotation(name_, codename_),
+                arr_localities.annotation_, name_, codename_));
+
+        return primitive_argument_type(result, attached_annotation);
+    }
```
Comment (hkaiser) [inline src/plugins/dist_matrixops/dist_diag.cpp:432]: Could you extract the various cases into separate functions, please? This function is very large and difficult to read.

File: src/plugins/matrixops/diag_operation.cpp [hunk 5]
```diff
@@ -81,41 +75,6 @@ namespace phylanx { namespace execution_tree { namespace primitives
...
-            break;
-        }
-
-        HPX_THROW_EXCEPTION(hpx::bad_parameter,
-            "diag_operation::diag0d",
-            generate_error_message(
-                "the diag primitive requires for all arguments to "
-                    "be numeric data types"));
-    }
-
```
Comment (hkaiser) [inline src/plugins/matrixops/diag_operation.cpp:118]: What's the rationale to remove `diag` for scalars? Is this change related to `diag_d`?

File: src/plugins/dist_matrixops/dist_diag.cpp (hunk 1 above)
Comment (NanmiaoWu) [reply src/plugins/dist_matrixops/dist_diag.cpp:30]: will delete it

File: src/plugins/dist_matrixops/dist_diag.cpp (hunk 2 above)
Comment (NanmiaoWu) [reply src/plugins/dist_matrixops/dist_diag.cpp:20]: Will delete it

File: src/plugins/dist_matrixops/dist_diag.cpp (hunk 3 above)
Comment (NanmiaoWu) [reply src/plugins/dist_matrixops/dist_diag.cpp]: Already put them in a new hpp file, called idx_tile_calculation_helper.hpp

File: src/plugins/matrixops/diag_operation.cpp (hunk 5 above)
Comment (NanmiaoWu) [reply src/plugins/matrixops/diag_operation.cpp:118]: Because Numpy diag does not consider scalar case. To comply with it, I delete the diag0d case.

File: src/plugins/dist_matrixops/dist_diag.cpp (hunk 4 above)
Comment (NanmiaoWu) [reply src/plugins/dist_matrixops/dist_diag.cpp:375]: Will do it.

**Comment by hkaiser** [comment]
@NanmiaoWu could you please rebase this on master to fix the build system issues?

**Comment by NanmiaoWu** [comment]
> @NanmiaoWu could you please rebase this on master to fix the build system issues?

Yes. Already done it.

File: phylanx/plugins/dist_matrixops/idx_tile_calculation_helper.hpp [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+#if !defined(PHYLANX_DIST_MATRIXOPS_IDX_TILE_CALCULATION_HELPER)
+#define PHYLANX_DIST_MATRIXOPS_IDX_TILE_CALCULATION_HELPER
+
+#include <hpx/assertion.hpp>
+
+#include <cstddef>
+#include <cstdint>
+#include <string>
+#include <tuple>
```
Comment (taless474) [inline phylanx/plugins/dist_matrixops/idx_tile_calculation_helper.hpp:16]: I think tuple is not needed here

Comment (NanmiaoWu) [reply phylanx/plugins/dist_matrixops/idx_tile_calculation_helper.hpp:16]: Yes. Already deleted it.

File: src/plugins/dist_matrixops/dist_diag.cpp [hunk 7]
```diff
@@ -0,0 +1,512 @@
...
+                    options are `row` or `column` tiling.
+                tile_index (int, optional): the tile index we need to generate
+                    the diag array for. A non-negative integer. If not
+                    given, it sets to current locality.
+                numtiles (int, optional): number of tiles of the returned array.
+                    If not given it sets to the number of localities in the
+                    application.
+            Returns:
+
+            A 1-D array of its k-th diagonal when a is a 2-D array; a 2-D array
```
Comment (taless474) [inline src/plugins/dist_matrixops/dist_diag.cpp:72]: Here we don't have dist_diag2d, so I guess we can remove "A 1-D array of its k-th diagonal when a is a 2-D array".

File: src/plugins/dist_matrixops/dist_diag.cpp [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+                    if (tile_idx >= numtiles)
+                    {
+                        HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                            "dist_diag::eval",
+                            this_->generate_error_message(
+                                "invalid tile index. Tile indices start from 0 "
+                                "and should be smaller than number of tiles"));
+                    }
+
+
```
Comment (taless474) [inline src/plugins/dist_matrixops/dist_diag.cpp:503]: The user might supply `a` with any dimension rather than 1. Would you please throw an exception in the case that `a`'s dimension is not 1?

File: src/plugins/dist_matrixops/dist_diag.cpp (hunk 7 above)
Comment (NanmiaoWu) [reply src/plugins/dist_matrixops/dist_diag.cpp:72]: Yes. Will delete it.

File: src/plugins/dist_matrixops/dist_diag.cpp (hunk 8 above)
Comment (NanmiaoWu) [reply src/plugins/dist_matrixops/dist_diag.cpp:502]: I am not sure I understand. Do you mean arr when mentioning a? And, why arr's dimension has to be 1? Or do you mean that it should be larger or equal to 1?

Comment (taless474) [reply src/plugins/dist_matrixops/dist_diag.cpp:502]: I mean if the user asks for the following (a 2-d array) what would be the result:
```
diag_d(annotate_d([[1,2,3],[4,5,6]],"name", list("tile",list("columns",0,3),list("rows",0,2))))
```

Comment (NanmiaoWu) [reply src/plugins/dist_matrixops/dist_diag.cpp:502]: ahh got you.because we currently only consider 1d array as input.
